outlines/models/llamacpp.py

```diff
@@ -1,304 +1,185 @@
-        elif input_ids.ndim == 1:
-            seq_tensor = input_ids.view(1, -1)[:, self.n_past :]
-        else:
-            raise Exception("Only one and two dimensional inputs allowed.")
-
-        tokens_total = torch.numel(seq_tensor[seq_tensor != self.pad_token_id])
-        batch = self.llama_instance.llama_batch_init(tokens_total, 0, 1)
-
-        seq_token_ids = []
-        for seq_idx, seq in enumerate(seq_tensor):
-            for token_pos, token_id in enumerate(seq):
-                if token_id == self.pad_token_id:
-                    break
-                batch.token[batch.n_tokens] = token_id.item()
-                batch.pos[batch.n_tokens] = token_pos
-                batch.seq_id[batch.n_tokens][0] = seq_idx
-                batch.n_seq_id[batch.n_tokens] = 1
-                batch.logits[batch.n_tokens] = False
-
-                batch.n_tokens += 1
-                self.n_past += 1
-
-            batch.logits[batch.n_tokens - 1] = True
-            seq_token_ids.append(batch.n_tokens - 1)
+        from llama_cpp import Llama
 
-        if self.llama_instance.llama_decode(self.ctx, batch) != 0:
-            print("Error decoding")
+        self.model = Llama(model_path, **kwargs)
+        self.logits_processor = logits_processor
+        self.tokenizer = LlamaCppTokenizer(self.model)
 
-        all_logits = []
-        for seq_token in seq_token_ids:
-            logits = self.llama_instance.llama_get_logits_ith(self.ctx, seq_token)
-            logits_list = (ctypes.c_float * self.n_vocab)(
-                *[logits[token_id] for token_id in range(self.n_vocab)]
-            )
-            logits_tensor = torch.tensor(logits_list)
-            all_logits.append(logits_tensor)
+    def __call__(
+        self,
+        prompts: Union[str, List[str]],
+        max_tokens: Optional[int] = None,
+        stop_at: Optional[Union[str, List[str]]] = None,
+        rng: Optional[torch.Generator] = None,
+        **model_kwargs,
+    ) -> Union[str, List[str]]:
+        from llama_cpp import LogitsProcessorList
+
+        if isinstance(prompts, str):
+            prompts = [prompts]
+
+        if rng is None:
+            rng = torch.Generator(device="cpu")
+            rng.seed()
+
+        results = []
+        for prompt in prompts:
+            processors = []
+            if self.logits_processor is not None:
+                processors = [copy.copy(self.logits_processor)]
+
+            result = self.model.create_completion(
+                prompt,
+                max_tokens=max_tokens,
+                stop=stop_at,
+                seed=rng.initial_seed(),
+                logits_processor=LogitsProcessorList(processors),
+                **model_kwargs,
+            )["choices"][0]["text"]
+            results.append(result)
+
+            self.model.reset()
+
+        formatted = [self.format_sequence(sequence) for sequence in results]
+
+        return formatted if len(formatted) > 1 else formatted[0]
```

Comment:
nit:

IMO, `List[prompt]` should always correspond to `List[generation]`. I think we should condition on the passed `prompts` type rather than the length of `formatted`.